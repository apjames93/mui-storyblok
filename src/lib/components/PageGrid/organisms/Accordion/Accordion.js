import React, {
  useState,
  lazy,
  Suspense,
} from 'react';
import PropTypes from 'prop-types';
import { Accordion as MuiAccordion } from '@material-ui/core';
import { dimensionProp, validComponentsRequired } from 'lib/utils/customProps';
import Storyblok from 'lib/utils/Storyblok';

const AccordionSummary = lazy(() => import('./components/AccordionSummary/AccordionSummary'));
const AccordionDetails = lazy(() => import('./components/AccordionDetails/AccordionDetails'));

const Accordion = ({
  rootClass,
  accordionDetails,
  accordionSummary,
  defaultExpanded,
  width,
  dataBlokC,
  dataBlokUid,
  storyblokClass,
}) => {
  const styles = Storyblok.arrayToMuiStyles(rootClass, { width });
  const [expanded, setExpanded] = useState(defaultExpanded);
  const handleChange = () => setExpanded(!expanded);

  const panelSummary = accordionSummary[0];
  const panelDetails = accordionDetails[0];
  return (
    <MuiAccordion
      expanded={expanded}
      onChange={handleChange}
      className={`${styles.root} ${storyblokClass}`}
      data-blok-c={dataBlokC}
      data-blok-uid={dataBlokUid}
    >
      <Suspense fallback={<></>}>
        <AccordionSummary {...panelSummary} />
      </Suspense>
      <Suspense fallback={<></>}>
        <AccordionDetails {...panelDetails} />
      </Suspense>
    </MuiAccordion>
  );
};


export default Accordion;

Accordion.propTypes = {
  /** stroyblok multiselect of css classes */
  rootClass: PropTypes.arrayOf(PropTypes.string),
  /** if true pannale is expanded on load */
  defaultExpanded: PropTypes.bool,
  /** width of panel */
  width(props, propName, componentName) {
    return dimensionProp(props, propName, componentName);
  },
  /** AccordionSummary Allowed maximum: 1 */
  accordionSummary(props, propName, componentName) {
    const components = ['AccordionSummary'];
    return validComponentsRequired(props, propName, componentName, components, 1);
  },
  /** AccordionDetails Allowed maximum: 1 */
  accordionDetails(props, propName, componentName) {
    const components = ['AccordionDetails'];
    return validComponentsRequired(props, propName, componentName, components, 1);
  },
};

Accordion.defaultProps = {
  width: '100%',
  defaultExpanded: false,
  rootClass: [],
  accordionSummary: [],
  accordionDetails: [],
};
