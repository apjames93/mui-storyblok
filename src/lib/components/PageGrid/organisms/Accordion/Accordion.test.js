import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import Accordion from './Accordion';

const accordionSummary = component => [{
  _editable: '<!--#storyblok#{"name": "AccordionSummary", "space": "48408", "uid": "947be9f0-47c3-4315-a95a-550f0c560eb5", "id": "307934"}-->',
  component: 'AccordionSummary',
  expandIcon: [{
    _editable: '<!--#storyblok#{"name": "Icon", "space": "48408", "uid": "947be9f0-47c3-4315-a95a-550f0c560eb5", "id": "307934"}-->',
    component: 'Icon',
    iconName: 'android',
  }],
  content: [{
    _editable: '<!--#storyblok#{"name": "Typography", "space": "48408", "uid": "947be9f0-47c3-4315-a95a-550f0c560eb5", "id": "307934"}-->',
    component: 'Typography',
    content: [{
      _editable: '<!--#storyblok#{"name": "Span", "space": "48408", "uid": "947be9f0-47c3-4315-a95a-550f0c560eb5", "id": "307934"}-->',
      component: 'Span',
      content: 'Span',
    }],
  }],
}];

function setup(summaryComp = true, accordionComp = 'AccordionSummary') {
  const props = {
    _editable: '<!--#storyblok#{"name": "Accordion", "space": "48408", "uid": "947be9f0-47c3-4315-a95a-550f0c560eb5", "id": "307934"}-->',
    accordionSummary: summaryComp ? [...accordionSummary(accordionComp)] : undefined,
    accordionDetails: [
      {
        _editable: '<!--#storyblok#{"name": "AccordionDetails", "space": "48408", "uid": "947be9f0-47c3-4315-a95a-550f0c560eb5", "id": "307934"}-->',
        component: 'AccordionDetails',
        content: [
          {
            _editable: '<!--#storyblok#{"name": "Typography", "space": "48408", "uid": "947be9f0-47c3-4315-a95a-550f0c560eb5", "id": "307934"}-->',
            component: 'Typography',
            content: [
              {
                _editable: '<!--#storyblok#{"name": "Span", "space": "48408", "uid": "947be9f0-47c3-4315-a95a-550f0c560eb5", "id": "307934"}-->',
                component: 'Span',
                content: 'text',
              },
            ],
          },
        ],
      },
    ],
  };
  const comp = shallow(<Accordion {...props} />);
  return { comp, props };
}

describe('<Accordion />', () => {
  const originalConsoleError = global.console.error;
  let warningMsg;

  beforeEach(() => {
    global.console.error = (...args) => {
      const propTypeFailures = [/Failed prop type/, /Warning: Recieved/];

      if (propTypeFailures.some(p => p.test(args[0]))) {
        warningMsg = [args[0]];
      }

      originalConsoleError(...args);
    };
  });

  it('renders Accordion', () => {
    const { comp } = setup();
    expect(comp).toBeDefined();
  });

  test('snapshot', () => {
    const { props } = setup();
    const tree = renderer.create(<Accordion {...props} />);
    expect(tree).toMatchSnapshot();
  });

  it('should handle change and toggle the expansion panel', () => {
    const { comp } = setup();
    const { expanded } = comp.find('WithStyles(ForwardRef(Accordion))').first().props();
    expect(expanded).toEqual(false);
    comp.find('WithStyles(ForwardRef(Accordion))').first().prop('onChange')();
    expect(comp.find('WithStyles(ForwardRef(Accordion))').first().props().expanded).toEqual(true);
  });

  it.skip('should give proper warning for propTypes on AccordionSummary if component is not passed down.', () => {
    setup(false);
    const expected = 'Warning: Failed prop type: Accordion: accordionSummary is required to have a length of 1 but recived length of 0\n    in Accordion (at Accordion.test.js:38)';
    expect(warningMsg[0]).toEqual(expected);
  });

  it.skip('should give proper warning for propTypes on AccordionSummary if component is invalid.', () => {
    setup(true, 'invalidComp');
    const expected = 'Warning: Failed prop type: Accordion: accordionSummary is required to have a length of 1 but recived length of 0\n    in Accordion (at Accordion.test.js:38)';
    expect(warningMsg[0]).toEqual(expected);
  });
});
