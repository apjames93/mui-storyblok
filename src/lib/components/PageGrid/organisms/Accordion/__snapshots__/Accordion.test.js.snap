// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Accordion /> snapshot 1`] = `
<div
  className="MuiPaper-root MuiAccordion-root makeStyles-root-3  MuiAccordion-rounded MuiPaper-elevation1 MuiPaper-rounded"
  data-blok-c=""
  data-blok-uid=""
>
  <div
    className="MuiCollapse-container MuiCollapse-hidden"
    style={
      Object {
        "minHeight": "0px",
      }
    }
  >
    <div
      className="MuiCollapse-wrapper"
    >
      <div
        className="MuiCollapse-wrapperInner"
      >
        <div
          role="region"
        />
      </div>
    </div>
  </div>
</div>
`;
